---
title: "TE Biogas: metatranscriptome analyses"
author: "Domenico Simone"
date: "May 18, 2020"
output:
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_knit$set(root.dir = "~/Dropbox/Research/slu/SLUBI_support/SLUBI_Project_TE_Biogas_snic2019-30-23")
knitr::opts_chunk$set(echo = TRUE, eval = FALSE, message = FALSE, warning = FALSE, echo=FALSE)
library(DESeq2)
library(tximport)
#library(GenomicFeatures)
library(tidyverse)
library(readr)
library(ggplot2)
library(gplots)
library(pheatmap)
library(sva)
library(plotly)
library(pheatmap)
library(RColorBrewer)
```

```{r, eval=TRUE, echo=FALSE}
## Read metadata
metadata <- read.table("Metatranscriptome/metadata.tsv", header = TRUE)
```

# Overview of gene expression at the bin level

Which bins are mostly expressed in the two conditions? We will integrate this info with phylogenetic placement.

```{r, eval=FALSE, echo=FALSE}
# Do it by calculating means across conditions

read_tsv("Metatranscriptome/Merged_Quant.sf") %>%  # read merged file from salmon (gene expr in TPM)
  rename(Gene=Name) %>% # rename "Name" column into "Gene"
  separate(col=Gene, c("Bin", NA), sep = "_[01]", remove = TRUE) %>% # add column with bin name (from gene name)
  group_by(Bin) %>% summarise_all(funs(sum)) %>% ungroup() %>% # summarise counts at bin level
  gather("Sample", "Abundance", 2:ncol(.)) %>% left_join(metadata) %>% select(-Replicate) %>% select(-Sample) %>%
  group_by_at(vars(Bin, Condition)) %>% summarise_all(funs(mean, sd)) %>% ungroup() %>% 
  #mutate(panel=nrow)  head()
  ggplot(aes(x=Bin, y=mean, fill=Condition)) + # plot with ggplot!
    geom_bar(stat="identity", position=position_dodge()) +
    geom_errorbar(aes(ymin=mean-sd, ymax=mean+sd), width=.2,
                 position=position_dodge(.9))

#ggplot(data=df, aes(x=reorder(player, df$is.winner=="TRUE", sum), fill=is.winner)) +

 #   facet_wrap(~Bin)
```

```{r, eval=FALSE, echo=FALSE}
# Plot as heatmap with ggplot. For rmarkdown report, I have used the plotly plot after this one

p <- read_tsv("Metatranscriptome/Merged_Quant.sf") %>%  # read merged file from salmon (gene expr in TPM)
  rename(Gene=Name) %>% # rename "Name" column into "Gene"
  separate(col=Gene, c("Bin", NA), sep = "_[01]", remove = TRUE) %>% # add column with bin name (from gene name)
  group_by(Bin) %>% summarise_all(funs(sum)) %>% ungroup() %>% # summarise counts at bin level
  gather("Sample", "Abundance", 2:ncol(.)) %>% left_join(metadata) %>% # create data frame for ggplot2
  ggplot(aes(x=Sample, y=Bin, fill=Abundance)) + # plot with ggplot!
    geom_tile() +
    #theme(axis.title.y = element_text(margin = margin(t = 0, r = 30, b = 30, l = 0))) +
    facet_grid(~Condition, scales="free")# %>%

# p
# gp <- ggplotly(p) #, tooltip="text")
  #layout(margin = list(l = 150))
 
# gp[['x']][['layout']][['annotations']][[2]][['x']] <- -0.13
# gp[['x']][['layout']][['annotations']][[1]][['y']] <- -0.05
# gp %>% layout(margin = list(l = 250))
```

The following heatmap shows expression levels (in TPMs) grouped by bin. You can hover each cell of the heatmap to get more info.

```{r, eval=TRUE, echo=FALSE, fig.height=20}
M1 <- read_tsv("Metatranscriptome/Merged_Quant.sf") %>%  # read merged file from salmon (gene expr in TPM)
  rename(Gene=Name) %>% # rename "Name" column into "Gene"
  separate(col=Gene, c("Bin", NA), sep = "_[01]", remove = TRUE) %>% # add column with bin name (from gene name)
  group_by(Bin) %>% summarise_all(funs(sum)) %>% ungroup()

Bins <- M1$Bin
M1 <- select(M1, -Bin) %>% as.matrix()

x <- list(
  title = "Sample"
#  titlefont = f
)
y <- list(
  title = "Bin"
#  titlefont = f
)

fig <- plot_ly(
    x = colnames(M1), y = Bins,
    z = M1, type = "heatmap"
) %>% layout(xaxis = x, yaxis = y)

fig
```

# Differential gene expression

Aims:

- Perform DGE
- Group by bin
- Group by taxonomy
- Group by functional annotation

## Methods

Use DESeq2 on salmon output.

```{r, echo=FALSE, eval=TRUE}
# Create data needed throughout the analysis
## Since salmon provides transcript-level estimates only, we need a two-column data frame linking transcript id (column 1) to gene id (column 2).
quant.files <- Sys.glob("Metatranscriptome/salmon/*/quant.sf")
tx2gene <-  quant.files %>%
            map(read_tsv) %>%    # read in all the files individually, using
                                 # the function read_csv() from the readr package
            reduce(rbind) %>%    # reduce with rbind into one dataframe
            transmute(transcript.id=Name,
                      gene.id=Name)

## Read counts into tximport
files <- file.path("Metatranscriptome/salmon", metadata$Sample, "quant.sf")
names(files) <- paste0("sample", 1:6)
#all(file.exists(files))
se <- tximport(files, type = "salmon", tx2gene = tx2gene)

## Create DESeq2 object
dds <- DESeqDataSetFromTximport(txi = se, colData = metadata, design = ~ Condition)
```

## Exploratory analysis and visualization

Prefilter the dataset (for visualization purposes **only**). Raw count data, how many genes?

```{r, echo=FALSE, eval=TRUE}
nrow(dds)
```

We can keep rows (= genes) with more than 1 count across all samples. How many genes do we have now?

```{r, echo=FALSE, eval=TRUE}
keep <- rowSums(counts(dds)) > 1
dds.exp <- dds[keep,]
nrow(dds.exp)
```

### Pick a method to visualise sample relationships

Source: https://www.bioconductor.org/packages/devel/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#exploratory-analysis-and-visualization

We need to choose between transformation methods for running exploratory analyses.

**Scatterplot of transformed counts from two samples.**

```{r, eval=TRUE}
show.transform <- function(dds.exp){
vsd <- vst(dds.exp, blind = FALSE)
rld <- rlog(dds.exp, blind = FALSE)
dds <- estimateSizeFactors(dds.exp)
df <- bind_rows(
as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
mutate(transformation = "log2(x + 1)"),
as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"),
as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"))
colnames(df)[1:2] <- c("x", "y")
lvls <- c("log2(x + 1)", "vst", "rlog")
df$transformation <- factor(df$transformation, levels=lvls)
ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
coord_fixed() + facet_grid( . ~ transformation)
}
show.transform(dds.exp)
```

**Which samples are similar to each other, which are different?** Does this fit to the expectation from the experiment’s design? We use the vst-transformed results to:

- draw a heatmap of sample-sample distances
- run a PCA analysis.

**Heatmap of sample-to-sample distances using the variance stabilizing transformed values.**

```{r, echo=FALSE, eval=TRUE}
vsd <- vst(dds.exp, blind = FALSE)

# We use the R function *dist* to calculate the Euclidean distance between samples. To ensure we have a roughly equal contribution from all genes, we use it on the VST data. We need to transpose the matrix of values using t, because the dist function expects the different samples to be rows of its argument, and different dimensions (here, genes) to be columns.
sampleDists <- dist(t(assay(vsd)))
# In order to plot the sample distance matrix with the rows/columns arranged by the distances in our distance matrix, we manually provide sampleDists to the clustering_distance argument of the pheatmap function. Otherwise the pheatmap function would assume that the matrix contains the data values themselves, and would calculate distances between the rows/columns of the distance matrix, which is not desired. We also manually specify a blue color palette using the colorRampPalette function from the RColorBrewer package.
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- dds$Sample
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
clustering_distance_rows = sampleDists,
clustering_distance_cols = sampleDists,
col = colors)
```

**PCA analysis**

```{r, echo=FALSE, eval=TRUE}
plotPCA.ggplot <- function(vsd){
  PCA.data <- plotPCA(vsd, intgroup = c("Condition"), returnData = TRUE)
  ggplot(PCA.data, aes(x = PC1, y = PC2, color = Condition)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", round(100 * attr(PCA.data, "percentVar"))[1], "% variance")) +
  #xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC1: ", round(100 * attr(PCA.data, "percentVar"))[2], "% variance")) +
  coord_fixed()
}

plotPCA.ggplot(vsd)
```

**Heatmap showing how much each gene deviates in a specific sample from the gene’s average across all samples.** Top 20 genes are shown.

```{r, echo=FALSE, eval=TRUE, fig.width=15, fig.height=15}
topVarGenes <- head(order(rowVars(assay(vsd)), decreasing = TRUE), 20)
mat  <- assay(vsd)[ topVarGenes, ]
mat  <- mat - rowMeans(mat)
colnames(mat) <- dds$Sample
anno <- as.data.frame(colData(vsd)[, c("Condition")])
colnames(anno) <- "Condition"
rownames(anno) <- dds$Sample
pheatmap(mat, annotation_col = anno)
```
